### ISSUE TITLE
Issue with type converters and dependency injection

### ISSUE THREAD

**Issue by sniffdk** [leekelleher/umbraco-ditto#63]
Hi guys

Not even sure if this has anything to do with Ditto, but here goes.

Been testing Ditto on a new project where I use Autofac to do some basic dependency injection.
However, I found, that as soon as I set the AutofacDependencyResolver as the current DependencyResolver type converters will simply stop running. Disabling DI will make everything work again. I've zip'ed a complete solution if you guys wan't to take a look -> https://dl.dropboxusercontent.com/u/3371537/NestedDittoDemo.zip

There is a single page using the Nested Content data type which again contains a multinodetreepicker. If you look in App_Code/Start.cs you can see the Autofac configuration I use. Take a look at Views/Page.cshtml for the code that is using Nested Content and Ditto.

Hopefully it's fairly straight forward to see whats going on, if not, let me know :)

Cheers
Mads

**Comment by leekelleher** [comment]
Hi @sniffdk - thanks for the demo site, it had helped a lot with trying to figure this out.

---

OK, firstly something to rule out - as it's misleading...

Ditto can only map to concrete class (meaning those that can be `new`'d).  Mapping to abstracts or interfaces wont work and will return null. (e.g. no mapping directly to `IPublishedContent` or `PublishedContentModel`)

It causes confusion, as the following property as a type mismatch...

```csharp
[TypeConverter(typeof(MultiNodeTreePickerConverter<SomeModel>))]
public IEnumerable<IPublishedContent> Members { get; set; }
```

... but that is due to type downcasting of `SomeModel` > `PublishedContentModel` > `IPublishedContent`.

---

Now for the real issue... :wink: 

In Ditto, when using a custom `TypeConverterAttribute` for advanced mapping, we make a call to [`DependencyResolver.Current.GetService()`](https://github.com/leekelleher/umbraco-ditto/blob/0.6.1/src/Our.Umbraco.Ditto/Extensions/PublishedContentExtensions.cs#L350), so to handle any Dependency Injection.

(_Personally, I don't do much DI work, but it was suggested in PR #20 - and sounded like a good idea at the time_)

Back to the demo site, when you set `AutofacDependencyResolver` - Ditto will try to use it to create an instance of the custom `TypeConverter`.  So when you registered the type...

```csharp
builder.RegisterType<MultiNodeTreePickerConverter<SomeModel>>();
```

... it worked fine.  Without it, it would return a null.

---

Current workaround is that you'll have to register all custom TypeConverters with Autofac :neutral_face:

Moving forwards, we should have a discussion on what is the expected behaviour of Ditto.  Should we first try to call `DependencyResolver.GetService()` and if it returns null, then fallback on the usual `Activator.CreateInstance()`?

// Open to comments - @JimBobSquarePants @mattbrailsford @stevetemple

**Comment by sniffdk** [comment]
Hi Lee

Yeah, that was the conclusion I arrived at as well. 
And not just register all type converters, but all type converters for each of your models.
Or, at least the different combinations you actually use :smile: 

I'm not really sure how much of an edge case the combination of Ditto, DI and type converters are, but it might just not be feasible to fix atm., when there is a possible workaround however verbose it is ..

**Comment by JimBobSquarePants** [comment]
@leekelleher  That fallback sounds sane to me. It's a two line fix.  I was under the impression that internally `DependencyResolver.GetService()` was supposed to do that though? https://github.com/leekelleher/umbraco-ditto/pull/20#issuecomment-70871159

@sniffdk Just to expand on why Ditto cannot map to anything other than concrete classes. This isn't due to limitations within Ditto, It's due constraints within the C# language. You cannot create instances of abstract nor interface types, you can only cast to them.  

Internally Ditto creates a new instance of the given type to assign property values to so anything other than a concrete type will fail. 

I think we can and should fix it for you internally. I want Ditto to fill your developer life with a warm fuzzy glow.

**Comment by leekelleher** [comment]
re: `DependencyResolver.GetService()` on [#20 (comment)](https://github.com/leekelleher/umbraco-ditto/pull/20#issuecomment-70871159) ... it will fallback on `Activator.CreateInstance` *if* there isn't a custom `DependencyResolver` configured. (At least that's what I found with my debugging)

I think we should add fallback code, it makes sense - as then if someone wants to handle the resolution in their own DI config, they can do... and if they don't, then it's the regular .NET way.

**Comment by JimBobSquarePants** [comment]
Agreed. I'll do the change at lunch. I want to change the logic slightly in the type initialization code anyhoo.

**Comment by JimBobSquarePants** [comment]
You wanna test that @leekelleher when you get a mo?

**Comment by sniffdk** [comment]
@JimBobSquarePants I'm perfectly aware of the constraints in C# regarding abstract and interface types :smiley: 
What I didn't initially understand was how the type converter concept works :sweat_smile: 
But thank you both for looking into the issue, you guys rock !

**Comment by JimBobSquarePants** [comment]
@sniffdk Sorry chum, no offence meant. Happy to help! :metal:

**Comment by sniffdk** [comment]
Non taken, just clarifying :smile:

**Comment by leekelleher** [comment]
@JimBobSquarePants Sure thing, I'll give it a test!

**Comment by leekelleher** [comment]
@sniffdk Quick question... when you use Autofac, do you noticed anything strange in the back-office?
During my testing, I found that certain application/trees don't load up. (I don't see anything in the trace-log though).

**Comment by leekelleher** [comment]
@JimBobSquarePants I've tried testing the latest in `develop` branch, but I'm hitting a YSoD with `DittoMultiNodeTreePickerConverter`.

Here's the YSoD...

![ditto-ysod](https://cloud.githubusercontent.com/assets/209066/7141206/f45e1a7a-e2c7-11e4-8a2b-f5a5e0cbb666.png)

It seems to be this line:

    var enumerable = value as IEnumerable<string>;

When I debug, the value of `value` is "1055,1056" (comma separated string). I can't see where it would be split?

Let me know if you need any more info.

**Comment by leekelleher** [comment]
@JimBobSquarePants - Ah, it's actually this line, [L105](https://github.com/leekelleher/umbraco-ditto/blob/develop/src/Our.Umbraco.Ditto/ComponentModel/TypeConverters/DittoMultiNodeTreePickerConverter.cs#L105) ...

    ((IEnumerable<int>)value).ToArray()

**Comment by leekelleher** [comment]
@JimBobSquarePants I commented out those lines (just for my test), and everything else works fine with the DI/Autofac enabled! :+1:

**Comment by sniffdk** [comment]
@leekelleher Yeah, you have to register at least api controllers in the Umbraco assembly, something like `builder.RegisterApiControllers(typeof(UmbracoApplication).Assembly);` , there is a bit more info here http://issues.umbraco.org/issue/U4-4181 . Also, I've read reports about people running into issues with the installer when upgrading Umbraco.

**Comment by JimBobSquarePants** [comment]
@leekelleher That's weird. 

What type is the property in the model? That code should only run if it is a generic type.

[#L64](https://github.com/leekelleher/umbraco-ditto/blob/develop/src/Our.Umbraco.Ditto/ComponentModel/TypeConverters/DittoMultiNodeTreePickerConverter.cs#L64)

**Comment by leekelleher** [comment]
@sniffdk - Cool, that worked.

---

@JimBobSquarePants - Yep, the value is definitely `"1055,1056"` (`string`).

Here's the POCO property...

```csharp
[TypeConverter(typeof(DittoMultiNodeTreePickerConverter))]
public IEnumerable<SomeModel> Members { get; set; }
```

So I guess since the POCO property type is `IEnumerable`, then `isGenericType` would be true?
The underlying Umbraco property-editor is regular "Umbraco.MultiNodeTreePicker".

Just to note, there are no custom PropertyValueConverters in play, just default Umbraco.

**Comment by JimBobSquarePants** [comment]
Urgh... I'm checking the wrong object. I should have been testing if the type of was generic.

I've uploaded changes so if you can check them, please do.

**Comment by leekelleher** [comment]
@JimBobSquarePants - Groovy... all A-OK! :smiley_cat:

@sniffdk - The custom TypeConverters in Ditto will now play nicely with Autofac (or other IoC containers)!  This will be in the next release v0.7.0 - or the latest build via [MyGet](https://www.myget.org/F/umbraco-ditto/).

**Comment by JimBobSquarePants** [comment]
That's freaking awesome!

**Comment by sniffdk** [comment]
That is indeed awesome !! :+1: :smile: 

@leekelleher glad to hear it worked.

**Comment by leekelleher** [comment]
Ace! I'm closing this issue. Fix will be in v0.7.0.
